### PR TITLE
Expose `cudaMemGetInfo`

### PR DESCRIPTION
**Pull request by coreylammie**
This PR resolves the second issue outlined in #58376, which has previously been discussed in #50722. 

`cudaMemGetInfo` is bound/exposed to the Python API. An example function call is provided below:

```
device_free, device_total = torch.cuda.mem_get_info(torch.device('cuda:0'))
print(device_free, device_total)
```

In  `CUDACachingAllocator.cpp`, in constant to my initial PR, the newly defined function `std::pair<size_t, size_t> raw_cuda_mem_get_info(int device)` has been moved from the `CUDACaching` namespace to the `cuda` namespace. In addition, as suugested by @ezyang, `det` has been removed from all function names.

### CODE REVIEW

File: c10/cuda/CUDACachingAllocator.cpp [hunk 1]
```diff
@@ -1504,5 +1504,13 @@ void raw_delete(void* ptr) {
 
 } // namespace CUDACachingAllocator
 
+std::pair<size_t, size_t> raw_cuda_mem_get_info(int device) {
+  C10_CUDA_CHECK(cudaGetDevice(&device));
+  size_t device_free;
+  size_t device_total;
+  C10_CUDA_CHECK(cudaMemGetInfo(&device_free, &device_total));
+  return {device_free, device_total};
+}
```
Comment (ezyang) [inline c10/cuda/CUDACachingAllocator.cpp:1513]: nit: There isn't really any reason to put this in CUDACachingAllocator.cpp, it doesn't need any access to the internal allocator, and really doesn't have anything to do with the caching allocator at all

**Comment by ezyang** [comment]
Sorry about not suggesting this originally, but I think the C portions of this patch can be done in one line, by modifying torch/csrc/cuda/shared/cudart.cpp

File: c10/cuda/CUDACachingAllocator.cpp (hunk 1 above)
Comment (coreylammie) [reply c10/cuda/CUDACachingAllocator.cpp]: Good point! I have re-defined this logic in `torch/csrc/cuda/shared/cudart.cpp`, as suggested. Please see: [https://github.com/pytorch/pytorch/pull/58635/commits/7d5c98a0306f23ad438302f29ab8fa321c316bcb](https://github.com/pytorch/pytorch/pull/58635/commits/7d5c98a0306f23ad438302f29ab8fa321c316bcb).

**Comment by coreylammie** [comment]
> Sorry about not suggesting this originally, but I think the C portions of this patch can be done in one line, by modifying torch/csrc/cuda/shared/cudart.cpp

I have moved all C++ logic/code from `c10.cuda.CUDACachingAllocator.cpp`, `c10.cuda.CUDACachingAllocator.h`, and `torch.csrc.cuda.Module.cpp` to `torch.csrc.cuda.shared.cudart.cpp` in [https://github.com/pytorch/pytorch/pull/58635/commits/7d5c98a0306f23ad438302f29ab8fa321c316bcb](https://github.com/pytorch/pytorch/pull/58635/commits/7d5c98a0306f23ad438302f29ab8fa321c316bcb). It may not be necessary, but I have included `<c10/cuda/CUDAException.h>` in `torch.csrc.cuda.shared.cudart.cpp`, so that `C10_CUDA_CHECK` can be called around `cudaGetDevice`.

**Review by ezyang — APPROVED** [review]

**Comment by vadimkantorov** [comment]
@coreylammie It seems it wasn't added to rst files and is invisible in docs: https://pytorch.org/docs/master/search.html?q=mem_get_info&check_keywords=yes&area=default

**Pull request closed without merging**